## Read PDFs

### Read Danish Rental Law

In [5]:
from langchain_community.document_loaders import PyPDFLoader
from langchain.text_splitter import CharacterTextSplitter
import pprint
import re
import numpy as np

In [6]:
file_path = "../src/data/lejeloven_2025.pdf"

loader = PyPDFLoader(
    file_path,)

documents = loader.load()

print("Number of pages:", len(documents))

Number of pages: 52


In [10]:
pprint.pp(documents[15].metadata)
pprint.pp(documents[15].page_content)

{'producer': 'PDFsharp 1.50.4000-netstandard '
             '(https://github.com/ststeiger/PdfSharpCore) (Original: Antenna '
             'House PDF Output Library 7.0.1600)',
 'creator': 'AH Formatter V7.0 MR3 for Windows (x64) : 7.0.4.45923 '
            '(2020-07-14T10:31+09)',
 'creationdate': '2023-02-17T11:06:05+01:00',
 'moddate': '2025-09-01T19:15:54+02:00',
 'title': 'Lov om leje',
 'trapped': '/False',
 'source': '../src/data/lejeloven_2025.pdf',
 'total_pages': 52,
 'page': 15,
 'page_label': '16'}
('§ 47. Reglerne i § 46 gælder også, hvis der pålægges ejendommen nye eller '
 'forøgede afgifter til vand, \n'
 'el, renovation, wc, skorstensfejning el.lign. efter takster, der er fastsat '
 'eller godkendt af det offentlige.\n'
 'Stk. 2.  Reglerne i § 46 gælder også, hvis der pålægges ejendommen nye vej- '
 'eller kloakbidrag eller \n'
 'lignende bidrag til det offentlige. Er bidrag pålagt ejendommen som en '
 'engangsydelse, kan udlejeren, når \n'
 'ydelsen er betalt, opkræve

Read as one pdf and extract page number by regex

In [7]:
loader = PyPDFLoader(
    file_path,
    mode="single",
    pages_delimiter="LOV nr 341 af 22/03/2022"
)
single_document = loader.load()

print("Number of pages:", len(single_document))

Number of pages: 1


In [8]:
single_document[0].metadata

{'producer': 'PDFsharp 1.50.4000-netstandard (https://github.com/ststeiger/PdfSharpCore) (Original: Antenna House PDF Output Library 7.0.1600)',
 'creator': 'AH Formatter V7.0 MR3 for Windows (x64) : 7.0.4.45923 (2020-07-14T10:31+09)',
 'creationdate': '2023-02-17T11:06:05+01:00',
 'moddate': '2025-09-01T19:15:54+02:00',
 'title': 'Lov om leje',
 'trapped': '/False',
 'source': '../docs/lejeloven_2025.pdf',
 'total_pages': 52}

In [9]:
pprint.pp(single_document[0].page_content[:8000])

('Udskriftsdato:\xa01.\xa0september\xa02025\n'
 'LOV\xa0nr\xa0341\xa0af\xa022/03/2022\xa0(Gældende)\n'
 'Lov\xa0om\xa0leje\n'
 'Ministerium:\xa0Social\xad\xa0og\xa0Boligministeriet Journalnummer:\xa0'
 'Indenrigs\xad\xa0og\xa0Boligmin.,\xa0j.nr.\xa02021\xad2261\n'
 'Senere\xa0ændringer\xa0til\xa0forskriften\n'
 'LOV\xa0nr\xa01311\xa0af\xa027/09/2022\xa0§\xa01\xa0\xa0\xad\xa0LOV\xa0nr\xa0'
 '482\xa0af\xa012/05/2023\xa0§\xa06\xa0\xa0\xad\xa0LOV\xa0nr\xa0753\xa0af\xa0'
 '13/06/2023\xa0§\xa09\xa0\xa0\xad\xa0\n'
 'LOV\xa0nr\xa01790\xa0af\xa028/12/2023\xa0§\xa04\xa0\xa0\xad\xa0LOV\xa0nr\xa0'
 '1793\xa0af\xa028/12/2023\xa0§\xa01LOV nr 341 af 22/03/2022Lov om leje\n'
 'VI MARGRETHE DEN ANDEN, af Guds Nåde Danmarks Dronning, gør vitterligt:\n'
 'Folketinget har vedtaget og Vi ved V ort samtykke stadfæstet følgende lov:\n'
 'Kapitel 1\n'
 'Anvendelsesområde m.v.\n'
 'Anvendelsesområde\n'
 '§ 1. Loven gælder for leje, herunder fremleje, af hus eller rum, uanset om '
 'lejeren er en person eller e

### Split single documents by chapters

In [10]:
# Check splitting by chapter
expected_chapters = re.findall(r"Kapitel \d{1,2}\n", single_document[0].page_content)

expected_chapters

['Kapitel 1\n',
 'Kapitel 2\n',
 'Kapitel 3\n',
 'Kapitel 4\n',
 'Kapitel 5\n',
 'Kapitel 6\n',
 'Kapitel 7\n',
 'Kapitel 8\n',
 'Kapitel 9\n',
 'Kapitel 10\n',
 'Kapitel 11\n',
 'Kapitel 12\n',
 'Kapitel 13\n',
 'Kapitel 14\n',
 'Kapitel 15\n',
 'Kapitel 16\n',
 'Kapitel 17\n',
 'Kapitel 18\n',
 'Kapitel 19\n',
 'Kapitel 20\n',
 'Kapitel 21\n',
 'Kapitel 22\n',
 'Kapitel 23\n',
 'Kapitel 24\n',
 'Kapitel 25\n',
 'Kapitel 26\n',
 'Kapitel 27\n',
 'Kapitel 28\n',
 'Kapitel 29\n']

In [11]:
chapter_splitter = CharacterTextSplitter(
    separator=r"Kapitel \d{1,2}\n",
    is_separator_regex=True,
    keep_separator=True,
)

chapter_documents = chapter_splitter.split_documents(single_document)

print(len(chapter_documents))

Created a chunk of size 7368, which is longer than the specified 4000
Created a chunk of size 6991, which is longer than the specified 4000
Created a chunk of size 31495, which is longer than the specified 4000
Created a chunk of size 13206, which is longer than the specified 4000
Created a chunk of size 4825, which is longer than the specified 4000
Created a chunk of size 12850, which is longer than the specified 4000
Created a chunk of size 4334, which is longer than the specified 4000
Created a chunk of size 5114, which is longer than the specified 4000
Created a chunk of size 5779, which is longer than the specified 4000
Created a chunk of size 4166, which is longer than the specified 4000
Created a chunk of size 8606, which is longer than the specified 4000
Created a chunk of size 11165, which is longer than the specified 4000
Created a chunk of size 10612, which is longer than the specified 4000
Created a chunk of size 9819, which is longer than the specified 4000
Created a chunk

27


In [12]:
found_chapters = [chapter.page_content.split('\n')[0] for chapter in chapter_documents]

found_chapters

['Udskriftsdato:\xa01.\xa0september\xa02025',
 'Kapitel 1',
 'Kapitel 2',
 'Kapitel 3',
 'Kapitel 4',
 'Kapitel 5',
 'Kapitel 6',
 'Kapitel 7',
 'Kapitel 8',
 'Kapitel 9',
 'Kapitel 10',
 'Kapitel 11',
 'Kapitel 12',
 'Kapitel 13',
 'Kapitel 14',
 'Kapitel 15',
 'Kapitel 16',
 'Kapitel 17',
 'Kapitel 18',
 'Kapitel 19',
 'Kapitel 20',
 'Kapitel 21',
 'Kapitel 22',
 'Kapitel 23',
 'Kapitel 24',
 'Kapitel 25',
 'Kapitel 29']

In [13]:

[print(chapter.replace("\n", "")) for chapter in expected_chapters if chapter.replace("\n", "") not in found_chapters]

Kapitel 26
Kapitel 27
Kapitel 28


[None, None, None]

In [14]:
pprint.pp(chapter_documents[found_chapters.index("Kapitel 25")].page_content)

('Kapitel 25\n'
 'Boligretten\n'
 'Saglig kompetence\n'
 '§ 202.  Tvister om lejeforhold, der er omfattet af denne lov, kan i 1. '
 'instans indbringes for byretten, \n'
 'hvis spørgsmålet ikke efter denne lov kan indbringes for huslejenævnet eller '
 'i Københavns Kommune \n'
 'ankenævnet. Retten benævnes boligretten.\n'
 'Stk. 2. Parterne kan dog, når der er opstået en tvist, aftale, at tvisten '
 'kan indbringes for boligretten, uden \n'
 'at huslejenævnet og i Københavns Kommune ankenævnet har behandlet sagen.\n'
 'Stk. 3. Stk. 1 og 2 begrænser ikke fogedrettens adgang til at gennemføre en '
 'umiddelbar fogedforretning, \n'
 'jf. retsplejelovens kapitel 55, om udsættelse af et lejemål, der er ophævet '
 'som følge af lejerens tilsidesæt-\n'
 'telse af god skik og orden.\n'
 'Fravigelighed\n'
 '§ 203. Dette kapitel kan ikke fraviges.\n'
 'Kapitel 26\n'
 'Regulering af beløb og beløbsgrænser\n'
 '§ 204. De i § 19, stk. 2 og 5, § 23, stk. 4, § 105, stk. 1, § 106, stk. 1, § '
 '117, s

In [15]:
pprint.pp(chapter_documents[found_chapters.index("Kapitel 3")].page_content)

('Kapitel 3\n'
 'Omkostningsbestemt husleje m.v.\n'
 'Huslejefastsættelse ved lejeaftalens indgåelse\n'
 '§ 19. Ved lejeaftalens indgåelse må lejen ikke fastsættes til et beløb, som '
 'overstiger det beløb, der kan \n'
 'dække ejendommens nødvendige driftsudgifter, jf. § 24, og afkastet af '
 'ejendommens værdi, jf. § 25. For \n'
 'lejemål, som er forbedret, kan der til lejen efter 1. pkt. lægges en '
 'beregnet forbedringsforhøjelse, jf. dog \n'
 'stk. 2.\n'
 'Stk. 2. Ved lejeaftalens indgåelse må lejen for lejemål, som er '
 'gennemgribende forbedret, ikke fastsæt-\n'
 'tes til et beløb, der overstiger det lejedes værdi efter § 42, stk. 2 og 3, '
 'jf. dog stk. 3-6, § 21, stk. \n'
 '1, og § 161, stk. 3. Ved lejemål, som er forbedret gennemgribende, forstås '
 'lejemål, hvor forbedringer \n'
 'efter principperne i § 128 væsentligt har forøget det lejedes værdi, og hvor '
 'forbedringsudgiften enten \n'
 'overstiger 2.280 kr. pr. m² eller et samlet beløb på 260.738 kr. '
 'Forbedringe

It seems like the character splitter from Langchain for some reason doesn't find chapter 26-28. Lets try splitting with regex and the inputting into documents instead

In [16]:
from langchain.schema import Document

def split_into_chapters(single_document, regex_pattern):
    text = single_document[0].page_content
    splits = re.split(regex_pattern, text)  # Capturing group includes the separator

    chunks = []
    for i in range(1, len(splits), 2):
        title = splits[i]
        content = splits[i+1] if i+1 < len(splits) else ""
        chunk = title + content
        chunks.append(Document(page_content=chunk, metadata={"chapter_title": title.strip()}))
    
    return chunks

chapter_regex = r"(Kapitel \d{1,2}\n)"
chapter_chunks = split_into_chapters(single_document, chapter_regex)
print(len(chapter_chunks))

29


In [17]:
found_chapters = [chapter.metadata["chapter_title"] for chapter in chapter_chunks]

[print(chapter.replace("\n", "")) for chapter in expected_chapters if chapter.replace("\n", "") not in found_chapters]

[]

In [18]:
pprint.pp(chapter_chunks[0].page_content[:500])

('Kapitel 1\n'
 'Anvendelsesområde m.v.\n'
 'Anvendelsesområde\n'
 '§ 1. Loven gælder for leje, herunder fremleje, af hus eller rum, uanset om '
 'lejeren er en person eller en \n'
 'virksomhed m.v. (juridisk person).\n'
 'Stk. 2. Loven gælder, selv om lejen skal betales med andet end penge, '
 'herunder ved arbejde.\n'
 'Stk. 3.  Bortset fra § 17 gælder loven dog ikke for aftaler om leje af bolig '
 'med fuld kost, for aftaler \n'
 'mellem et hotel og dets gæster og for lejeforhold om beboelseslejligheder og '
 'andre beboelsesrum, \n'
 'herunder somm')


It seems like this splitting worked as intended. Renaming chapter_chunks to chapter_documents for clarity.

In [19]:
chapter_documents = chapter_chunks

### Split chapters to paragraphs

In [20]:
re.findall(r"\n§ \d{1,3}", single_document[0].page_content)

['\n§ 1',
 '\n§ 2',
 '\n§ 3',
 '\n§ 4',
 '\n§ 5',
 '\n§ 6',
 '\n§ 7',
 '\n§ 8',
 '\n§ 9',
 '\n§ 10',
 '\n§ 11',
 '\n§ 26',
 '\n§ 12',
 '\n§ 13',
 '\n§ 14',
 '\n§ 15',
 '\n§ 16',
 '\n§ 17',
 '\n§ 18',
 '\n§ 19',
 '\n§ 20',
 '\n§ 21',
 '\n§ 22',
 '\n§ 23',
 '\n§ 24',
 '\n§ 25',
 '\n§ 2',
 '\n§ 26',
 '\n§ 27',
 '\n§ 28',
 '\n§ 29',
 '\n§ 30',
 '\n§ 31',
 '\n§ 32',
 '\n§ 23',
 '\n§ 33',
 '\n§ 34',
 '\n§ 35',
 '\n§ 36',
 '\n§ 37',
 '\n§ 38',
 '\n§ 39',
 '\n§ 40',
 '\n§ 41',
 '\n§ 42',
 '\n§ 43',
 '\n§ 44',
 '\n§ 45',
 '\n§ 46',
 '\n§ 48',
 '\n§ 49',
 '\n§ 50',
 '\n§ 51',
 '\n§ 52',
 '\n§ 53',
 '\n§ 54',
 '\n§ 55',
 '\n§ 56',
 '\n§ 57',
 '\n§ 58',
 '\n§ 59',
 '\n§ 60',
 '\n§ 182',
 '\n§ 61',
 '\n§ 62',
 '\n§ 63',
 '\n§ 64',
 '\n§ 65',
 '\n§ 66',
 '\n§ 67',
 '\n§ 68',
 '\n§ 69',
 '\n§ 70',
 '\n§ 71',
 '\n§ 72',
 '\n§ 73',
 '\n§ 74',
 '\n§ 75',
 '\n§ 77',
 '\n§ 78',
 '\n§ 79',
 '\n§ 80',
 '\n§ 81',
 '\n§ 82',
 '\n§ 83',
 '\n§ 84',
 '\n§ 85',
 '\n§ 86',
 '\n§ 87',
 '\n§ 88',
 '\n§ 89',
 '\n§ 90

Seems to be that this expression catches some unwanted splits, for example paragraph 26 under paragraph 11. Need to specify more

In [21]:
paragraph_regex = r"((?<=[\w\.])\n§ \d{1,3}\.)"

expected_paragraphs = re.findall(paragraph_regex, single_document[0].page_content)
expected_paragraphs = [re.findall(r"\d{1,3}", para)[0] for para in expected_paragraphs]
print(len(expected_paragraphs))
print(expected_paragraphs[:5])
print("Max paragraph:", np.max([int(num) for num in expected_paragraphs]))


210
['1', '2', '3', '4', '5']
Max paragraph: 213


The lenght and the maximum number paragraphs is correct when looking in the pdf.

In [22]:
def split_into_paragraphs(chapter, regex_pattern):
    text = chapter.page_content
    splits = re.split(regex_pattern, text)
    # splits: [before first paragraph, para_heading1, para_content1, para_heading2, para_content2, ...]
    paragraphs = []
    for i in range(1, len(splits), 2):
        para_heading = splits[i]
        para_content = splits[i+1] if i+1 < len(splits) else ""
        chunk = para_heading + para_content
        paragraph_number = int(re.findall(r"\d{1,3}", para_heading)[0])
        paragraphs.append(
            Document(
                page_content=chunk.strip(),
                metadata={
                    "chapter_title": chapter.metadata.get("chapter_title"),
                    "paragraph_number": paragraph_number
                }
            )
        )
    return paragraphs

paragraph_documents = [split_into_paragraphs(chapter, paragraph_regex) for chapter in chapter_documents]
paragraph_documents = [para for sublist in paragraph_documents for para in sublist]

In [23]:
paragraph_documents[:3]

[Document(metadata={'chapter_title': 'Kapitel 1', 'paragraph_number': 1}, page_content='§ 1. Loven gælder for leje, herunder fremleje, af hus eller rum, uanset om lejeren er en person eller en \nvirksomhed m.v. (juridisk person).\nStk. 2. Loven gælder, selv om lejen skal betales med andet end penge, herunder ved arbejde.\nStk. 3.  Bortset fra § 17 gælder loven dog ikke for aftaler om leje af bolig med fuld kost, for aftaler \nmellem et hotel og dets gæster og for lejeforhold om beboelseslejligheder og andre beboelsesrum, \nherunder sommerhuse, kolonihavehuse og andre fritidsboliger, som er udlejet til ferie- og fritidsmæssige \nformål.\nStk. 4. Loven gælder endvidere for leje af ustøttede private plejeboliger, jf. dog § 1, stk. 4, i erhvervsle-\njeloven. §§ 7 og 9, kapitel 3, § 53, stk. 2, §§ 105-107, 109, 113, 119-121, 123, § 127, stk. 3, § 135 og \nkapitel 24 gælder dog ikke for ustøttede private plejeboliger.\nStk. 5.  Ustøttede private plejeboliger, jf. stk. 4, er tidligere institu

In [24]:
pprint.pp(paragraph_documents[0].metadata)
pprint.pp(paragraph_documents[0].page_content[:500])

{'chapter_title': 'Kapitel 1', 'paragraph_number': 1}
('§ 1. Loven gælder for leje, herunder fremleje, af hus eller rum, uanset om '
 'lejeren er en person eller en \n'
 'virksomhed m.v. (juridisk person).\n'
 'Stk. 2. Loven gælder, selv om lejen skal betales med andet end penge, '
 'herunder ved arbejde.\n'
 'Stk. 3.  Bortset fra § 17 gælder loven dog ikke for aftaler om leje af bolig '
 'med fuld kost, for aftaler \n'
 'mellem et hotel og dets gæster og for lejeforhold om beboelseslejligheder og '
 'andre beboelsesrum, \n'
 'herunder sommerhuse, kolonihavehuse og andre fritidsboliger, som')


There's also sub-divisions of each chapter which would nice to include as metadata.

![image.png](attachment:image.png)


In [25]:
pprint.pp(chapter_documents[0].page_content)

('Kapitel 1\n'
 'Anvendelsesområde m.v.\n'
 'Anvendelsesområde\n'
 '§ 1. Loven gælder for leje, herunder fremleje, af hus eller rum, uanset om '
 'lejeren er en person eller en \n'
 'virksomhed m.v. (juridisk person).\n'
 'Stk. 2. Loven gælder, selv om lejen skal betales med andet end penge, '
 'herunder ved arbejde.\n'
 'Stk. 3.  Bortset fra § 17 gælder loven dog ikke for aftaler om leje af bolig '
 'med fuld kost, for aftaler \n'
 'mellem et hotel og dets gæster og for lejeforhold om beboelseslejligheder og '
 'andre beboelsesrum, \n'
 'herunder sommerhuse, kolonihavehuse og andre fritidsboliger, som er udlejet '
 'til ferie- og fritidsmæssige \n'
 'formål.\n'
 'Stk. 4. Loven gælder endvidere for leje af ustøttede private plejeboliger, '
 'jf. dog § 1, stk. 4, i erhvervsle-\n'
 'jeloven. §§ 7 og 9, kapitel 3, § 53, stk. 2, §§ 105-107, 109, 113, 119-121, '
 '123, § 127, stk. 3, § 135 og \n'
 'kapitel 24 gælder dog ikke for ustøttede private plejeboliger.\n'
 'Stk. 5.  Ustøttede privat

Cannot figure out a way of how to extract the headings for now, revisit this later.

### Extract page - paragraph relation

In [26]:
pprint.pp(documents[1].page_content[:3000])

('Lov om leje\n'
 'VI MARGRETHE DEN ANDEN, af Guds Nåde Danmarks Dronning, gør vitterligt:\n'
 'Folketinget har vedtaget og Vi ved V ort samtykke stadfæstet følgende lov:\n'
 'Kapitel 1\n'
 'Anvendelsesområde m.v.\n'
 'Anvendelsesområde\n'
 '§ 1. Loven gælder for leje, herunder fremleje, af hus eller rum, uanset om '
 'lejeren er en person eller en \n'
 'virksomhed m.v. (juridisk person).\n'
 'Stk. 2. Loven gælder, selv om lejen skal betales med andet end penge, '
 'herunder ved arbejde.\n'
 'Stk. 3.  Bortset fra § 17 gælder loven dog ikke for aftaler om leje af bolig '
 'med fuld kost, for aftaler \n'
 'mellem et hotel og dets gæster og for lejeforhold om beboelseslejligheder og '
 'andre beboelsesrum, \n'
 'herunder sommerhuse, kolonihavehuse og andre fritidsboliger, som er udlejet '
 'til ferie- og fritidsmæssige \n'
 'formål.\n'
 'Stk. 4. Loven gælder endvidere for leje af ustøttede private plejeboliger, '
 'jf. dog § 1, stk. 4, i erhvervsle-\n'
 'jeloven. §§ 7 og 9, kapitel 3, § 5

In [27]:
def extract_paragraphs_from_page(doc, regex_pattern):
    paragraphs = re.findall(regex_pattern, doc.page_content)

    page_paragraph = {int(re.findall(r"\d{1,3}", para)[0]):doc.metadata["page"]  for para in paragraphs}
    return page_paragraph


paragraph_pages = [extract_paragraphs_from_page(doc, paragraph_regex) for doc in documents]
paragraph_pages = {k:v for d in paragraph_pages for k,v in d.items()} # unnest list
paragraph_pages

{1: 1,
 2: 1,
 3: 1,
 4: 1,
 5: 2,
 6: 2,
 7: 2,
 8: 2,
 9: 2,
 10: 3,
 11: 3,
 12: 3,
 13: 4,
 14: 4,
 15: 4,
 16: 4,
 17: 5,
 18: 5,
 19: 5,
 20: 6,
 21: 7,
 22: 7,
 23: 7,
 24: 8,
 25: 8,
 26: 9,
 27: 10,
 28: 10,
 29: 10,
 30: 10,
 31: 11,
 32: 11,
 33: 11,
 34: 11,
 35: 11,
 36: 11,
 37: 12,
 38: 12,
 39: 12,
 40: 12,
 41: 12,
 42: 13,
 43: 13,
 44: 14,
 45: 14,
 46: 14,
 48: 15,
 49: 15,
 50: 16,
 51: 16,
 52: 16,
 53: 16,
 54: 16,
 55: 16,
 56: 17,
 57: 17,
 58: 17,
 59: 17,
 60: 17,
 61: 18,
 62: 18,
 63: 18,
 64: 18,
 65: 18,
 66: 18,
 67: 18,
 68: 19,
 69: 19,
 70: 19,
 71: 19,
 72: 19,
 73: 20,
 74: 20,
 75: 20,
 77: 21,
 78: 21,
 79: 21,
 80: 21,
 81: 21,
 82: 21,
 83: 21,
 84: 22,
 85: 22,
 86: 22,
 87: 22,
 88: 23,
 89: 23,
 90: 23,
 91: 23,
 92: 23,
 93: 23,
 94: 23,
 95: 24,
 96: 24,
 97: 24,
 98: 24,
 99: 24,
 100: 24,
 101: 25,
 102: 25,
 103: 25,
 104: 25,
 105: 25,
 106: 25,
 107: 26,
 108: 26,
 109: 26,
 110: 26,
 111: 26,
 112: 26,
 113: 27,
 114: 27,
 115: 27,
 1

In [28]:
def add_page_to_paragraphs(paragraph_docs, paragraph_page_map):
    for para_doc in paragraph_docs:
        para_num = para_doc.metadata["paragraph_number"]
        para_doc.metadata["page"] = paragraph_page_map.get(para_num)
    return paragraph_docs

paragraph_documents = add_page_to_paragraphs(paragraph_documents, paragraph_pages)
paragraph_documents[:3]

[Document(metadata={'chapter_title': 'Kapitel 1', 'paragraph_number': 1, 'page': 1}, page_content='§ 1. Loven gælder for leje, herunder fremleje, af hus eller rum, uanset om lejeren er en person eller en \nvirksomhed m.v. (juridisk person).\nStk. 2. Loven gælder, selv om lejen skal betales med andet end penge, herunder ved arbejde.\nStk. 3.  Bortset fra § 17 gælder loven dog ikke for aftaler om leje af bolig med fuld kost, for aftaler \nmellem et hotel og dets gæster og for lejeforhold om beboelseslejligheder og andre beboelsesrum, \nherunder sommerhuse, kolonihavehuse og andre fritidsboliger, som er udlejet til ferie- og fritidsmæssige \nformål.\nStk. 4. Loven gælder endvidere for leje af ustøttede private plejeboliger, jf. dog § 1, stk. 4, i erhvervsle-\njeloven. §§ 7 og 9, kapitel 3, § 53, stk. 2, §§ 105-107, 109, 113, 119-121, 123, § 127, stk. 3, § 135 og \nkapitel 24 gælder dog ikke for ustøttede private plejeboliger.\nStk. 5.  Ustøttede private plejeboliger, jf. stk. 4, er tidlig

## Simple RAG

In [29]:
import os
from dotenv import load_dotenv
load_dotenv()
os.environ["LANGCHAIN_TRACING_V2"] = "true"
os.environ["LANGCHAIN_API_KEY"] = os.getenv("LANGCHAIN_API_KEY")
os.environ["LANGCHAIN_PROJECT"] = "rental-contract-rag"

In [30]:
from langchain_core.vectorstores import InMemoryVectorStore
from langchain.chat_models import ChatOpenAI
from langchain_core.runnables import RunnablePassthrough
from langchain_core.output_parsers import StrOutputParser
from langchain import hub


os.environ["OPENAI_API_KEY"] = os.getenv("OPENAI_API_KEY")

from langchain_openai import OpenAIEmbeddings

embeddings = OpenAIEmbeddings(model="text-embedding-3-small")

vector_store = InMemoryVectorStore.from_documents(
    embedding=embeddings,
    documents=paragraph_documents,
    collection_name="rental_contract_law_2025"
)

retriever = vector_store.as_retriever(search_type="similarity", search_kwargs={"k":10})

# Prompt
prompt = hub.pull("rlm/rag-prompt")

# LLM
llm = ChatOpenAI(model_name="gpt-4-turbo", temperature=0)

# Post-processing
def format_docs(docs):
    return "\n\n".join(doc.page_content for doc in docs)

# Chain
rag_chain = (
    {"context": retriever | format_docs, "question": RunnablePassthrough()}
    | prompt
    | llm
    | StrOutputParser()
)

# Question
answer = rag_chain.invoke("I hvilken pargraf kan jeg læse om tilbagebetaling af for meget betalt husleje?")
pprint.pp(answer)

C:\Users\MartinHallberg\AppData\Local\Temp\ipykernel_29800\1396238872.py:26: LangChainDeprecationWarning: The class `ChatOpenAI` was deprecated in LangChain 0.0.10 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-openai package and should be used instead. To use it run `pip install -U :class:`~langchain-openai` and import as `from :class:`~langchain_openai import ChatOpenAI``.
  llm = ChatOpenAI(model_name="gpt-4-turbo", temperature=0)


'Du kan læse om tilbagebetaling af for meget betalt husleje i § 40.'


Try an example paragraph from a contract and see if it can answer.

In [31]:
from langchain.prompts import ChatPromptTemplate


template = """You are a helpful assistant that answers questions about the Danish Rent Act based on excerpts from the legal text.
{context}

In the answers, refer to the specific paragraph numbers (e.g., § 1, § 2, etc.) and the page number where the information was found, in the following structure:

Answer: 'Your answer here'
References:
- Paragraphs: '§ X, § Y'
- Pages: 'Y'

If you cannot find the answer in the excerpts, say "I don't know".


Question: {question}
"""
specified_prompt = ChatPromptTemplate.from_template(template)
specified_prompt

ChatPromptTemplate(input_variables=['context', 'question'], input_types={}, partial_variables={}, messages=[HumanMessagePromptTemplate(prompt=PromptTemplate(input_variables=['context', 'question'], input_types={}, partial_variables={}, template='You are a helpful assistant that answers questions about the Danish Rent Act based on excerpts from the legal text.\n{context}\n\nIn the answers, refer to the specific paragraph numbers (e.g., § 1, § 2, etc.) and the page number where the information was found, in the following structure:\n\nAnswer: \'Your answer here\'\nReferences:\n- Paragraphs: \'§ X, § Y\'\n- Pages: \'Y\'\n\nIf you cannot find the answer in the excerpts, say "I don\'t know".\n\n\nQuestion: {question}\n'), additional_kwargs={})])

Check what answers we get for some questions:

**Example questions:**  
* Is it OK for the landlord to have 3 weeks' notice to start work in my apartment?
    * Answer should be 'No' and a referral to paragraph 103.

In [32]:
# Chain
rag_chain = (
    {"context": retriever | format_docs, "question": RunnablePassthrough()}
    | specified_prompt
    | llm
    | StrOutputParser()
)

# Question
answer = rag_chain.invoke("Er det OK at udlejer har 3 ugers varsel på at igangsætte arbejder i min lejlighed?")
pprint.pp(answer)

('Answer: Nej, det er ikke OK ifølge de oplysninger, der er givet. Udlejeren '
 'kan med 6 ugers varsel iværksætte arbejder i det lejede, når udførelsen ikke '
 'er til væsentlig ulempe for lejeren. For andre arbejder kræves der 3 '
 'måneders varsel, og kun uopsættelige reparationer kan foretages uden '
 'varsel.\n'
 'References:\n'
 '- Paragraphs: § 103\n'
 '- Pages: Not specified')


* Can the landloard have 1 month termination period of the contract?
    * Answer should be 'Only in if the rental is regarding a room, or in the same house as the landloard, paragraph 175."

In [33]:
answer = rag_chain.invoke("Kan en udlejer opsigle lejemålet med 1 måned varsel?")
pprint.pp(answer)

('Answer: Nej, en udlejer kan generelt ikke opsige et lejemål med kun 1 måneds '
 'varsel. Opsigelsesvarslet for udlejeren afhænger af typen af lejemål og de '
 'specifikke omstændigheder. For eksempel er opsigelsesvarslet 3 måneder til '
 'den første hverdag i en måned, der ikke er dagen før en helligdag, medmindre '
 'andet er aftalt, som det fremgår af § 175, stk. 1. Der er dog en undtagelse '
 'for visse typer lejemål, hvor opsigelsesvarslet er 1 måned, men dette gælder '
 'kun for de lejeforhold, der er nævnt i § 170, nr. 1.\n'
 '\n'
 'References:\n'
 "- Paragraphs: '§ 175'\n"
 "- Pages: 'N/A'")


* Can the landloard have a deposit of 5 months?
    * Answer should be 'No, maximum deposit is 3 months"

In [34]:
answer = rag_chain.invoke("Kan min udlejer have 5 måneders depositum?")
pprint.pp(answer)

('Answer: Nej, din udlejer kan ikke kræve 5 måneders depositum. Ifølge § 59 '
 'kan udlejeren ved lejemålets indgåelse kræve et depositum svarende til højst '
 '3 måneders leje.\n'
 '\n'
 'References:\n'
 '- Paragraphs: § 59\n'
 '- Pages: Not specified')


It seems like the model can locate relavant information from simple questions. There does however seem to be an issue to retrieve the page number from the metadata, might have to be added to the doc when retrieving.

## Read and index contract PDF

In [35]:
contract_file = "../docs/contract_template_with_info_printed.pdf"
contract_loader = PyPDFLoader(
    contract_file,
    mode="single",
)
contract_single = contract_loader.load()

print("Number of pages:", len(contract_single))

Number of pages: 1


In [36]:
pprint.pp(contract_single[0].metadata)
pprint.pp(contract_single[0].page_content[:3000])

{'producer': 'Microsoft: Print To PDF',
 'creator': 'PyPDF',
 'creationdate': '2025-09-08T17:00:55+02:00',
 'author': 'Martin Hallberg',
 'moddate': '2025-09-08T17:00:55+02:00',
 'title': 'LEJEKONTRAKT for beboelse.pdf',
 'source': '../docs/contract_template_with_info_printed.pdf',
 'total_pages': 14}
'\n\x0c\n\x0c\n\x0c\n\x0c\n\x0c\n\x0c\n\x0c\n\x0c\n\x0c\n\x0c\n\x0c\n\x0c\n\x0c'


It doesn't seem like `pypdf` recognizes the text that I have added in the template contract.

Try with other packages instead:

* `pdfplumber`
    * Did not work


In [37]:
import fitz
doc = fitz.open("../docs/contract_template_with_info_printed.pdf")
text = ""
for page in doc:
    text += page.get_text()

pprint.pp(text[:3000])

''


In [38]:
from pdf2image import convert_from_path
import pytesseract

pytesseract.pytesseract.tesseract_cmd = r"C:\Projects\tesseract\tesseract.exe"

pages = convert_from_path("../docs/contract_template_with_info_printed.pdf",
                          poppler_path=r"C:\Projects\poppler-25.07.0\Library\bin")
text = ""
for page in pages:
    text += pytesseract.image_to_string(page)

print(text[:3000])

Typeformular A, 10. udgave

Lejemalsnr.:

L EJ E KO N T RA KT for beboelse

Lejekontrakt til anvendelse i lejeaftaler om beboelseslejligheder,
herunder blandede lejemal, og verelser i private udlejningsejen-
domme.

En rakke bestemmelser i lejelovgivningen er ufravigelige, medens
andre gyldigt kan fraviges ved aftale. @nsker parterne at aftale
fravigelser af lejelovgivningens almindelige regler og/eller
lejekontrakten, skal det aftalte anfores i kontraktens § 11.

Afialte fravigelser ma ikke anfores direkte i kontraktteksten (ved
overstregning eller lign.), medmindre der i den fortrykte tekst er
givet serlig adgang hertil.

Enkelte vilkar i den fortrykte tekst er fremhevet med fed og kursiv.

Disse vilkar er fravigelser af lejelovgivningens almindelige regler.
Hvis parterne har aftalt de forhold, der er kursiveret i kontrakten, er

§ 1. Parterne og det lejede

Autoriseret af Indenrigs- og Boligministeriet den 1. september 2022.

det ikke nedvendigt at anfore de samme forhold i kontrakt

### Read and extract main contract info

In [ ]:
prompt_contract_main_info = """
You are a helpful assistant that summarizes rental contracts.
Extract the main information from the rental contract below. Focus on key details such as:
 - Parties involved (landlord and tenant)
 - Monthly rental amount and payment terms
 - Type of property and address
 - Duration of the lease and termination conditions
 - Deoposit amount and prepaid rent
 - What emenities are included (e.g., parking, dishwasher, etc.)
 - What utilities are included (e.g., heating, water, electricity, internet, etc.)
 - Renters responsibilities and obligations, specifically regarding inside and outside maintenance, repairs, and what they include,and any other duties outlined in the contract.

 Answer with in a json format with keys, as specified in example below. Do not add or change any key in the json.
 "landlord": "John Doe",
   "tenant": "Maria Smith",
   "monthly_rental_amount": "3000 DKK",
   "payment_terms": "First of each month",
   "property_type": "Rental apartment",
   "property_address": "Street 123, City, ZIP",
   "lease_duration": "12 months",
   "termination_conditions": "3 months notice",
   "deposit_amount": "6000 DKK",
   "prepaid_rent": "3000 DKK",
   "amenities": "Parking, Dishwasher",
   "utilities": "heating": "Included", "water": "Included", "electricity": "Included", "internet": "Included",
   "renters_responsibilities": 
       "inside_maintenance": "Tenant is responsible for all inside maintenance",
        "outside_maintenance": "Landlord is responsible for outside maintenance",
       "repairs": "Tenant must report all repairs needed"
   
 '

 {contract_text}
"""

from langchain.prompts import PromptTemplate

prompt_template = PromptTemplate(
    template=prompt_contract_main_info,
    input_variables=["contract_text"],
)
from langchain import LLMChain

llm = ChatOpenAI(model_name="gpt-4-turbo", temperature=0)

llm_chain = LLMChain(
    llm=llm,
    prompt=prompt_template,
)

answer = llm_chain.run(contract_text=text)

pprint.pp(answer)

('```json\n'
 '{\n'
 '  "landlord": "Martin Hallberg",\n'
 '  "tenant": "Martin Hallberg",\n'
 '  "monthly_rental_amount": "3000 DKK",\n'
 '  "payment_terms": "Due on the first of each month",\n'
 '  "property_type": "Apartment",\n'
 '  "property_address": "Nørrebro, Copenhagen",\n'
 '  "lease_duration": "Indefinite until terminated by notice",\n'
 '  "termination_conditions": "3 months notice by tenant, specific conditions '
 'for landlord as per §§ 170 and 171 in the rental law",\n'
 '  "deposit_amount": "9000 DKK",\n'
 '  "prepaid_rent": "9000 DKK",\n'
 '  "amenities": "Common laundry, bicycle parking, garage",\n'
 '  "utilities": {\n'
 '    "Heating": "Included",\n'
 '    "Water": "Included",\n'
 '    "Electricity": "Included",\n'
 '    "Internet": "Not mentioned"\n'
 '  },\n'
 '  "renters_responsibilities": {\n'
 '    "Inside maintenance": "Tenant is responsible for all inside maintenance '
 'including painting, whitening, wallpapering, and floor treatment",\n'
 '    "Outside main

### Split contract

Try recursive splitting

In [40]:
from langchain.text_splitter import RecursiveCharacterTextSplitter

text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=500,
    chunk_overlap=100,
    length_function=len,
    is_separator_regex=False,
)

contract_documents = text_splitter.create_documents([text])
print(len(contract_documents))

128


In [41]:

for i, doc in enumerate(contract_documents[:20]):
    print(f"""Document {i}----------------------------------------
          {doc.page_content}""")

Document 0----------------------------------------
          Typeformular A, 10. udgave

Lejemalsnr.:

L EJ E KO N T RA KT for beboelse

Lejekontrakt til anvendelse i lejeaftaler om beboelseslejligheder,
herunder blandede lejemal, og verelser i private udlejningsejen-
domme.

En rakke bestemmelser i lejelovgivningen er ufravigelige, medens
andre gyldigt kan fraviges ved aftale. @nsker parterne at aftale
fravigelser af lejelovgivningens almindelige regler og/eller
lejekontrakten, skal det aftalte anfores i kontraktens § 11.
Document 1----------------------------------------
          Afialte fravigelser ma ikke anfores direkte i kontraktteksten (ved
overstregning eller lign.), medmindre der i den fortrykte tekst er
givet serlig adgang hertil.

Enkelte vilkar i den fortrykte tekst er fremhevet med fed og kursiv.

Disse vilkar er fravigelser af lejelovgivningens almindelige regler.
Hvis parterne har aftalt de forhold, der er kursiveret i kontrakten, er

§ 1. Parterne og det lejede

Autori

### Find documents relevant for a question

In [46]:
from langchain.output_parsers import StructuredOutputParser, ResponseSchema

def analyze_contract_docs(contract_docs, llm_model=llm, prompt_template=prompt_template):

    # Define the expected schema for the output
    response_schemas = [
        ResponseSchema(name="summary", description="Concise summary of the contract excerpt"),
        ResponseSchema(name="to_be_checked", description="Flag if the excerpt should be checked by a legal expert (true/false)")
    ]
    output_parser = StructuredOutputParser.from_response_schemas(response_schemas)

    llm_chain = LLMChain(
        llm=llm_model,
        prompt=prompt_template,
        output_parser=output_parser
    )

    contract_info = {}

    for i, doc in enumerate(contract_docs):
        print(f"Processing document {i+1}/{len(contract_docs)}")
        try:
            answer = llm_chain.run(contract_text=doc.page_content)
            contract_info[i+1] = {
                "summary": answer["summary"],
                "to_be_checked": answer["to_be_checked"],
                "text": doc.page_content
            }
        except Exception as e:
            print(f"Error processing document {i+1}: {e}")
            contract_info[i+1] = {"error": str(e)}

    return contract_info


prompt_analyze_contract = """
You are a helpful assistant that specializes in understanding rental contracts.

You are tasked with two things:
    - Analyze the following excerpt from a rental contract and provide a concise summary of its content, max one sentence.
    - Flag if the excerpt contains information that should be manually checked by a legal expert, e.g., deposit, prepaid rent, termination conditions, access to the property, etc.

Contract excerpt: {contract_text}

Return your asnwer in a json format with keys:
 - summary
 - to_be_checked (true/false)
"""

prompt_template_analyze = PromptTemplate(
    template=prompt_analyze_contract,
    input_variables=["contract_text"],
)

contract_analysis = analyze_contract_docs(contract_documents[:10], prompt_template=prompt_template_analyze)
contract_analysis

Processing document 1/10
Processing document 2/10
Processing document 3/10
Processing document 4/10
Processing document 5/10
Processing document 6/10
Processing document 7/10
Processing document 8/10
Processing document 9/10
Processing document 10/10


{1: {'summary': 'This rental contract template is for residential leases, including mixed-use properties and rooms in private rental buildings, and specifies that some provisions of the lease law are mandatory while others can be modified by agreement in section 11 of the contract.',
  'to_be_checked': True,
  'text': 'Typeformular A, 10. udgave\n\nLejemalsnr.:\n\nL EJ E KO N T RA KT for beboelse\n\nLejekontrakt til anvendelse i lejeaftaler om beboelseslejligheder,\nherunder blandede lejemal, og verelser i private udlejningsejen-\ndomme.\n\nEn rakke bestemmelser i lejelovgivningen er ufravigelige, medens\nandre gyldigt kan fraviges ved aftale. @nsker parterne at aftale\nfravigelser af lejelovgivningens almindelige regler og/eller\nlejekontrakten, skal det aftalte anfores i kontraktens § 11.'},
 2: {'summary': 'The contract excerpt specifies that deviations from the standard text should not be made directly on the contract unless explicitly allowed, and certain terms highlighted in bold

It seems like it will be difficult to distinguish between general information and very important information, e.g., deposit, termination, responsibility etc.

It might therefore be a better way to explicitly find the most important information and use that in comparison with the law.

### Redo single extract from contract


In [62]:
from langchain.output_parsers import StructuredOutputParser, ResponseSchema


prompt_contract_all_info = """
You are a helpful assistant that summarizes rental contracts.
Extract the main information from the rental contract below. Focus on key details such as:
 - Parties involved (landlord and tenant)
 - Monthly rental amount and payment terms
 - Rental type (lejemål), ownership, andel, rental, subrental, room or other
 - Address
 - Start date of the lease
 - Duration of the lease and termination conditions
 - Deposit amount and prepaid rent
 - How and when rent can be adjusted
 - What amenities are included (e.g., parking, dishwasher, etc.)
 - What utilities are included (e.g., heating, water, electricity, internet, etc.)
 - Renters responsibilities and obligations, specifically regarding inside and outside maintenance, repairs, and what they include,and any other duties outlined in the contract.

 

 Answer with in a json format with keys, as specified in example below. Do not add or change any key in the json.
 "landlord": "John Doe",
   "tenant": "Maria Smith",
   "monthly_rental_amount": "3000 DKK",
   "payment_terms": "First of each month",
   "rental_type": "Ejerlejlighed",
   "property_address": "Street 123, City, ZIP",
   "lease_start_date": "2023-01-01",
   "lease_duration": "12 months",
   "termination_conditions": "3 months notice",
   "price_adjustments": "Rent can be adjusted annually based on CPI",
   "deposit_amount": "6000 DKK",
   "prepaid_rent": "3000 DKK",
   "amenities": "Parking, Dishwasher",
   "utilities": "heating": "Included", "water": "Included", "electricity": "Included", "internet": "Included",
   "renters_responsibilities": 
       "inside_maintenance": "Tenant is responsible for all inside maintenance",
        "outside_maintenance": "Landlord is responsible for outside maintenance",
       "repairs": "Tenant must report all repairs needed"

Always include all the above keys, even if it is not explicitly stated in the contract.

 {contract_text}
"""


# Define the expected schema for the output
response_schemas = [
    ResponseSchema(name="landlord", description="Name of the landlord"),
    ResponseSchema(name="tenant", description="Name of the tenant"),
    ResponseSchema(name="monthly_rental_amount", description="Monthly rental amount"),
    ResponseSchema(name="payment_terms", description="Payment terms"),
    ResponseSchema(name="rental_type", description="Type of property, ownership, rental, andel, single room, sublent or other"),
    ResponseSchema(name="property_address", description="Property address"),
    ResponseSchema(name="lease_start_date", description="Lease start date"),
    ResponseSchema(name="lease_duration", description="Duration of the lease"),
    ResponseSchema(name="termination_conditions", description="Termination conditions"),
    ResponseSchema(name="price_adjustments", description="Price adjustment conditions"),
    ResponseSchema(name="deposit_amount", description="Deposit amount"),
    ResponseSchema(name="prepaid_rent", description="Prepaid rent"),
    ResponseSchema(name="amenities", description="Amenities included"),
    ResponseSchema(name="utilities", description="Utilities included, and if they are included or not"),
    ResponseSchema(name="renters_responsibilities", description="Responsibilities of the renter, specifically regarding inside and outside maintenance, repairs, and any other duties outlined in the contract"),
]

output_parser = StructuredOutputParser.from_response_schemas(response_schemas)

prompt_template = PromptTemplate(
    template=prompt_contract_all_info,
    input_variables=["contract_text"],
    output_parser=output_parser,
)

llm_chain = LLMChain(
    llm=llm,
    prompt=prompt_template,
    output_parser=output_parser,
)

answer = llm_chain.run(contract_text=text)
pprint.pp(answer)

{'landlord': 'Martin Hallberg',
 'tenant': 'Martin Hallberg',
 'monthly_rental_amount': '3000 DKK',
 'payment_terms': 'First of each month',
 'rental_type': 'Subrental',
 'property_address': 'Nørrebro, København',
 'lease_start_date': '2025-11-01',
 'lease_duration': 'Indefinite until terminated',
 'termination_conditions': '3 months notice by tenant, varies for landlord as '
                           'per §§ 170 and 171 in the rental law',
 'price_adjustments': 'Rent can be adjusted annually based on the Danish '
                      'Statistics Consumer Price Index',
 'deposit_amount': '9000 DKK',
 'prepaid_rent': '9000 DKK',
 'amenities': 'Common laundry, bicycle parking, garage',
 'utilities': {'heating': 'Included',
               'water': 'Included',
               'electricity': 'Included',
               'internet': 'Not specified'},
 'renters_responsibilities': {'inside_maintenance': 'Tenant is responsible for '
                                                    'all inside

## Use RAG and LLM to validate contract

Termination conditions

In [65]:
termination_conditions = answer.get("termination_conditions", "Not specified")

termination_answer = rag_chain.invoke(f"Are these termination conditions correct? {termination_conditions}")
pprint.pp(termination_answer)

('Answer: Yes, the termination conditions you mentioned are generally correct. '
 'The tenant has a notice period of 3 months to the first weekday of a month '
 'that is not the day before a holiday. However, for leases mentioned in § '
 '170, no. 1, the notice period is 1 month. For the landlord, the notice '
 'period varies and can be at least 1 year for terminations under § 170, nos. '
 '2 and 3, and § 171, stk. 1, no. 1, to the first weekday in a month that is '
 'not the day before a holiday.\n'
 '\n'
 'References:\n'
 '- Paragraphs: § 175\n'
 '- Pages: 42')


Price adjustment

In [69]:
price_adjustment = answer.get("price_adjustments", "Not specified")

termination_answer = rag_chain.invoke(f"Are these price adjustment conditions correct? The price can be adjusted as the landloard seem fit")
pprint.pp(termination_answer)

('Answer: No, the price adjustments are not solely at the discretion of the '
 'landlord. According to § 42, the rent can be increased if it is '
 'significantly lower than the value of the rented property, but this increase '
 'must be reasonable and consider the usual rent in the area for similar '
 'properties. Additionally, § 23 allows for rent increases to cover necessary '
 "operational costs and property yield, but not beyond the property's value as "
 'assessed under § 42. There are also specific rules and conditions under '
 'which rent can be adjusted, including the requirement for written notice and '
 'comparisons to similar properties.\n'
 '\n'
 'References:\n'
 '- Paragraphs: § 42, § 23\n'
 '- Pages: Not specified in the query')


In [72]:
from langchain.chat_models import ChatOpenAI

llm = ChatOpenAI(model_name="gpt-4-turbo", temperature=0)

question = "In Denmark, can the tenant of an apartment be responsible for all maintenance of the building?", 
response = llm.invoke(question)
pprint.pp(response.content)


('In Denmark, the responsibility for maintenance in rental properties is '
 'typically divided between the tenant and the landlord. Generally, the '
 'landlord is responsible for the overall maintenance and repairs of the '
 'building, including structural aspects and major systems like heating, '
 'plumbing, and electrical systems. This includes ensuring that the building '
 'adheres to safety and health regulations.\n'
 '\n'
 'Tenants, on the other hand, are usually responsible for smaller maintenance '
 'tasks within their own apartment. This might include things like keeping the '
 'unit clean, making minor repairs, and reporting any issues that require more '
 'significant attention to the landlord. The specific responsibilities can '
 'vary based on the rental agreement, so it is important for tenants to review '
 'their lease documents to understand their obligations.\n'
 '\n'
 'It is uncommon and generally not legally supported for a tenant to be '
 'responsible for all mainten